numpy/random/_generator.pyx

```diff
@@ -873,11 +873,16 @@ cdef class Generator:
                 raise ValueError("a and p must have same size")
             p_sum = kahan_sum(pix, d)
             if np.isnan(p_sum):
-                raise ValueError("probabilities contain NaN")
+                raise ValueError("Probabilities contain NaN")
             if np.logical_or.reduce(p < 0):
-                raise ValueError("probabilities are not non-negative")
+                raise ValueError("Probabilities are not non-negative")
             if abs(p_sum - 1.) > atol:
-                raise ValueError("probabilities do not sum to 1")
+                raise ValueError("Probabilities do not sum to 1. "
+                                 "You can typically solve this issue with "
+                                 "`p = p / np.sum(p)`. "
+                                 "In rare cases this may not work due to round-off error, "
```

Comment:
Are there examples of normalization by `np.sum(p)` not working? The tolerance `atol = max(atol, np.sqrt(np.finfo(p.dtype).eps))` is pretty generous.

Reply:
Assuming that `p` came in as `float64`, yes. When the user provides a `float32` array that sums reasonably close to 1 in `float32` arithmetic, we're simply casting it to `float64`, and that's one of the more common sources of "spuriously" hitting this exception.

Reply:
Right. I was going to suggest that _if_ we were to provide a recommendation, it would be sufficient to recommend that the user normalize by `np.sum(p, dtype=np.float64)` rather than providing a primary strategy and backup strategy.